pkg/manifestclient/write_roundtripper.go

```diff
@@ -0,0 +1,133 @@
+package manifestclient
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/util/sets"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/server"
+	"net/http"
+	"sigs.k8s.io/yaml"
+)
+
+// Saves all mutations for later serialization and/or inspection.
+// In the case of updating the same thing multiple times, all mutations are stored and it's up to the caller to decide
+// what to do.
+type writeTrackingRoundTripper struct {
+	// requestInfoResolver is the same type constructed the same way as the kube-apiserver
+	requestInfoResolver *apirequest.RequestInfoFactory
+
+	actionTracker *AllActionsTracker
+}
+
+func newWriteRoundTripper() *writeTrackingRoundTripper {
+	return &writeTrackingRoundTripper{
+		actionTracker: &AllActionsTracker{},
+		requestInfoResolver: server.NewRequestInfoResolver(&server.Config{
+			LegacyAPIGroupPrefixes: sets.NewString(server.DefaultLegacyAPIPrefix),
+		}),
+	}
+}
+
+func (mrt *writeTrackingRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp := &http.Response{}
+
+	retJSONBytes, err := mrt.roundTrip(req)
+	if err != nil {
+		resp.StatusCode = http.StatusInternalServerError
+		resp.Status = http.StatusText(resp.StatusCode)
+		resp.Body = io.NopCloser(bytes.NewBufferString(err.Error()))
+		return resp, nil
+	}
+
+	resp.StatusCode = http.StatusOK
+	resp.Status = http.StatusText(resp.StatusCode)
+	resp.Body = io.NopCloser(bytes.NewReader(retJSONBytes))
+	// We always return application/json. Avoid clients expecting proto for built-ins.
+	// this may or may not work for apply.  Guess we'll find out.
+	resp.Header = make(http.Header)
+	resp.Header.Set("Content-Type", "application/json")
+
+	return resp, nil
+}
+
+func (mrt *writeTrackingRoundTripper) roundTrip(req *http.Request) ([]byte, error) {
+	requestInfo, err := mrt.requestInfoResolver.NewRequestInfo(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading requestInfo: %w", err)
+	}
+
+	if !requestInfo.IsResourceRequest {
+		return nil, fmt.Errorf("non-resource requests are not supported by this implementation")
+	}
+	if len(requestInfo.Subresource) != 0 && requestInfo.Subresource != "status" {
+		return nil, fmt.Errorf("subresource %v is not supported by this implementation", requestInfo.Subresource)
```

Comment:
why ? what about `scale` ?

Reply:
> why ? what about `scale` ?

If we need it, we'll add it.  For now getting this off the ground with enough to start using.